frontend/src/host_orchestrator/orchestrator/userartifacts.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orchestrator
+
+import (
+	apiv1 "github.com/google/android-cuttlefish/frontend/src/liboperator/api/v1"
+)
+
+// Abstraction for managing user artifacts for launching CVDs.
+type UserArtifactsManager interface {
+	NewToken() (*apiv1.UploadToken, error)
+}
+
+// Options for creating instances of UserArtifactsManager implementations.
+type UserArtifactsManagerOpts struct {
+	// The directory where to store the artifacts.
+	Dir string
+	// Factory of UUID values
+	UUIDFactory func() string
+}
+
+// An implementation of the UserArtifactsManager interface.
+type UserArtifactsManagerImpl struct {
+	UserArtifactsManagerOpts
+}
+
+// Creates a new instance of UserArtifactsManagerImpl.
+func NewUserArtifactsManagerImpl(opts UserArtifactsManagerOpts) UserArtifactsManager {
+	return &UserArtifactsManagerImpl{
+		UserArtifactsManagerOpts: opts,
+	}
+}
+
+func (m *UserArtifactsManagerImpl) NewToken() (*apiv1.UploadToken, error) {
+	if err := createDir(m.Dir, false); err != nil {
+		return nil, err
+	}
+	token := m.UUIDFactory()
+	if err := createDir(m.Dir+"/"+token, false); err != nil {
```

Comment:
Shouldn't this fail if the directory already exists?

Also, take a look at [Call site readability and boolean parameters](https://g3doc.corp.google.com/devtools/library_club/g3doc/totw/94.md?cl=head)

Reply:
good catch! done.

[Call site readability and boolean parameters](https://g3doc.corp.google.com/devtools/library_club/g3doc/totw/94.md?cl=head) was a great reading, will make the relevant changes afterwards.